google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/DirectPathFallbackIT.java

```diff
@@ -0,0 +1,250 @@
+
+    // Verify that the above check was meaningful, by verifying that the blackhole actually dropped
+    // packets.
+    assertThat(numBlocked.get()).isGreaterThan(0);
+
+    // Make sure that the client will start reading from IPv6 again by sending new requests and
+    // checking the injected IPv6 counter has been updated.
+    blackholeIPv6.set(false);
+    numIPv6Read.set(0);
+
+    while (numIPv6Read.get() < MIN_COMPLETE_READ_CALLS) {
+      for (int i = 0; i < NUM_RPCS_TO_SEND; i++) {
+        instrumentedClient.readRow(testEnvRule.env().getTableId(), "nonexistent-row");
+      }
+
+      Thread.sleep(100);
+    }
+
+    assertWithMessage("Failed to upgrade back to DirectPath")
+        .that(numIPv6Read.get())
+        .isGreaterThan(MIN_COMPLETE_READ_CALLS);
+  }
+
+  /**
+   * This is a giant hack to enable testing DirectPath CFE fallback.
+   *
+   * <p>It unwraps the {@link ComputeEngineChannelBuilder} to inject a NettyChannelHandler to signal
+   * IPv6 packet loss.
+   */
+  private void injectNettyChannelHandler(ManagedChannelBuilder<?> channelBuilder) {
+    try {
+      // Extract the delegate NettyChannelBuilder using reflection
+      Field delegateField = ComputeEngineChannelBuilder.class.getDeclaredField("delegate");
+      delegateField.setAccessible(true);
+
+      ComputeEngineChannelBuilder gceChannelBuilder =
+          ((ComputeEngineChannelBuilder) channelBuilder);
+      Object delegateChannelBuilder = delegateField.get(gceChannelBuilder);
+
+      NettyChannelBuilder nettyChannelBuilder = (NettyChannelBuilder) delegateChannelBuilder;
+      nettyChannelBuilder.channelFactory(channelFactory);
+      nettyChannelBuilder.eventLoopGroup(eventLoopGroup);
+    } catch (NoSuchFieldException | IllegalAccessException e) {
+      throw new RuntimeException("Failed to inject the netty ChannelHandler", e);
+    }
+  }
+
+  /** @see com.google.cloud.bigtable.data.v2.it.DirectPathFallbackIT.MyChannelHandler */
+  private class MyChannelFactory implements ChannelFactory<NioSocketChannel> {
+    @Override
+    public NioSocketChannel newChannel() {
+      NioSocketChannel channel = new NioSocketChannel();
+      channel.pipeline().addLast(new MyChannelHandler());
+
+      return channel;
+    }
+  }
+
+  /**
+   * A netty {@link io.grpc.netty.shaded.io.netty.channel.ChannelHandler} that can be instructed to
+   * make IPv6 packets disappear
+   */
+  private class MyChannelHandler extends ChannelDuplexHandler {
+    private boolean isIPv6;
+
+    @Override
+    public void connect(
+        ChannelHandlerContext ctx,
+        SocketAddress remoteAddress,
+        SocketAddress localAddress,
+        ChannelPromise promise)
+        throws Exception {
+
+      this.isIPv6 =
+          (remoteAddress instanceof InetSocketAddress)
+              && ((InetSocketAddress) remoteAddress).getAddress() instanceof Inet6Address;
+
+      if (!(isIPv6 && blackholeIPv6.get())) {
```

Comment:
nit - I'd extract `isIPv6 && blackholeIPv6.get()` into a variable to be more descriptive

Reply:
done